### ISSUE TITLE
sensor.solaredge_b1_available_energy becomes "Unknown" if capacity exceeds 9700

### ISSUE THREAD

**Issue by liquidv** [WillCodeForCats/solaredge-modbus-multi#230]
Before reporting a bug, make sure your configuration is correct, your LAN is working, and you have read about the "Known Issues" in the wiki:
https://github.com/WillCodeForCats/solaredge-modbus-multi/wiki

**Describe the bug**
sensor.solaredge_b1_available_energy becomes "Unknown" if capacity exceeds 9700

**Expected behavior**
If capacity exceeds 9700 it should show a value instead of "Unknown"

**Screenshots**
![image](https://user-images.githubusercontent.com/122752718/212568231-0d73d73c-336c-4dfc-bcfb-7941e3264112.png)

**Home Assistant (please complete the following information):**
 - Home Assistant Core Version:
Home Assistant 2023.1.2
Supervisor 2022.12.1
Operating System 9.4
Frontend 20230104.0 - latest

 - solaredge-modbus-multi Version: 2.2.9

**Comment by WillCodeForCats** [comment]
What's your value of b1 device attribute `batt_rated_energy`?

**Comment by WillCodeForCats** [comment]
I will probably also need to see debug logs for `B_Energy_Available` and `B_RatedEnergy` at the time the problem is occurring.

**Comment by WillCodeForCats** [comment]
`batt_rated_energy` is not a sensor, it's an attribute of the device sensor.

**Comment by WillCodeForCats** [comment]
Should be `sensor.solaredge_b1_device` and it will have a bunch of attributes which are static info about the device.

**Comment by liquidv** [comment]
Is this it? Sorry being dumb here...

![image](https://user-images.githubusercontent.com/122752718/212570009-4e4b9599-a850-470d-9039-4b3d7cad34aa.png)

**Comment by liquidv** [comment]
![image](https://user-images.githubusercontent.com/122752718/212570417-263aaef6-407e-4c28-93e3-ca49977d80c9.png)

**Comment by WillCodeForCats** [comment]
Yes, the second one. Maximum energy is correct, but available energy either is not returning a valid value while the battery is in charge mode, or it's reporting something out of range. The sensor for available energy will only show values that are between 0 and the batt rated energy, or 0-9700 in your case.

But I would need to see the debug log values to be sure which condition it's not meeting.

**Comment by liquidv** [comment]
I believe I have turned debug on, would you be able to let me know how to retreive those logs? Thanks

**Comment by liquidv** [comment]
Using the Solaredge API and various android apps to get battery info (SolarView and SolarMON), battery tips over the 9700 Wh range sometimes when fully charged

![image](https://user-images.githubusercontent.com/122752718/212571191-0113f3be-5d83-46a7-9d80-65b4d6b3b727.png)

**Comment by WillCodeForCats** [comment]
As this integration is currently written it validates the range as 0 to nameplate rating, and rejects values outside of the specs.

**Comment by liquidv** [comment]
Ok understood, I had max capacity issues with my battery which Solaredge had fixed and I had been monitoring it closely over the past month and still keeping an eye on it.
It's not an issue during charging as the value is there.

Any future updates to increase the value instead of matching nameplate rating?

**Comment by WillCodeForCats** [comment]
The max rated value is a number that we ask the battery to tell us when the integration starts up, it's not set in code.

**Comment by liquidv** [comment]
I see, thank you

**Comment by WillCodeForCats** [comment]
No values are altered or "normalized".

**Comment by WillCodeForCats** [comment]
You need to stop accusing me of saying I'm doing things that I'm not or saying I'm falsifying data, and reread to comprehend what I actually said is happening.

**Comment by WillCodeForCats** [comment]
> Ok understood, I had max capacity issues with my battery which Solaredge had fixed and I had been monitoring it closely over the past month and still keeping an eye on it.
> It's not an issue during charging as the value is there.

My question is, does it only then report a value that exceeds its reported maximum rating while in charge mode, or does it still do that in other modes too?

**Comment by liquidv** [comment]
> > Ok understood, I had max capacity issues with my battery which Solaredge had fixed and I had been monitoring it closely over the past month and still keeping an eye on it.
> > It's not an issue during charging as the value is there.
> 
> My question is, does it only then report a value that exceeds its reported maximum rating while in charge mode, or does it still do that in other modes too?

It does it in all modes regardless if it is charging or not (charging/discharging/idle).
It stays in an Unknown state until the battery is fully charged again and the BMS decides what value to assign it.
It has now finished charging and now states a value since it's below nameplate value

![image](https://user-images.githubusercontent.com/122752718/212574151-e2da180e-8222-43e4-bc1f-2098c39d61bf.png)

**Comment by WillCodeForCats** [comment]
My thought was if it's only happening during charging mode then I can add a condition, but if it happens in any mode then that won't work.

**Comment by WillCodeForCats** [comment]
So basically there has to be some kind of validation not just displaying any raw value. I wrote this integration because I came from using other integrations that did just show raw data, but it's common enough for the inverters to return bad data that causes errors in long term statistics or pushes graphs out of range for that time period which is something I don't want to deal with as a user. I know it's possible to go into the database and try to remove bad data, but I don't want to do that personally in my HA system so I don't expect anyone else to do that either.

Right now the data validation standard for Maximum and Available energy is to use the rated energy the battery tells us is what its limits are. These limits can be different for different makes and models. If there's an inaccuracy factor in the fixed limits vs. the dynamic data, say the BMS reports within 2% +/-, I could factor that in. But I don't know if that's a thing so I'm just speculating.

In your case the raw value is slightly above so probably fine to display, but there has to be a line somewhere. If I let it display anything then it's guaranteed I'll just get bug reports about values too high or too low messing up displays and we're back to having to validate again.

**Comment by rtlinux** [comment]
> Using the Solaredge API and various android apps to get battery info (SolarView and SolarMON), battery tips over the 9700 Wh range sometimes when fully charged

Typically the available energy drops to 9.5kWh, when installed with hardwired communication cable (RS485) . I am guessing you are probably using EnergyNet (wireless)?

I also noticed your SMCU version is 1.0.62 as opposed to the latest 1.0.66 - do you happen to know if your battery is an early version with the built-in fire extinguisher? The max energy higher than rated might have something to do with temperature coefficient. Perhaps 19.2C is the ideal temperature for this battery :)

**Comment by liquidv** [comment]
Not sure if you are on the solaredge facebook enthusiasts group but I had
endless troubles with my battery max capacity issue where it would only
give me approximately 7-8kwh battery opposed to 9.7kwh. It took SE 5 weeks
to diagnose the issue and fix it, prompting an issue with the voltage
during charge or something along those lines. Now I get the full 9.7kwh,
and I am also on hardwired ethernet opposed to wireless from the start of
installation.
Not sure if I have a built-in fire extinguisher, but my battery was fitted
in November 2022 so it's relatively new.
Temperature has little effect on the charge capacity after they fixed the
issue. I am still getting 9.7kwh even when the temp drops to 2c outside
where my battery resides in the unheated/uninsulated garage.

[image: image.png]

**Comment by liquidv** [comment]
Thank you for the explanation, it seemed like it could have been an easy
fix at the start but as you stated it's dependent on nameplate rating so
there isn't changeable.
I guess not everyone is too concerned about their max capacity rating.
happy to leave things as it stands. Thank you

**Comment by WillCodeForCats** [comment]
Do you have a record of the highest number over spec that you've seen reported?

**Comment by liquidv** [comment]
> Do you have a record of the highest number over spec that you've seen reported?

Mine is 9754, however my brother in laws who live 2 doors down the road drew the lucky straw and has a SE battery which seems to top out at 10,225! (see pic)

![image](https://user-images.githubusercontent.com/122752718/212737491-ea4579ee-0735-4010-b649-f1ffdda69d56.png)

**Comment by WillCodeForCats** [comment]
Do you have any that show over rated capacity not in charging state?

**Comment by liquidv** [comment]
> 

![image](https://user-images.githubusercontent.com/122752718/212742018-ab5187bb-e52f-4662-bb0c-2b5107ece448.png)

**Comment by liquidv** [comment]
> Do you have any that show over rated capacity not in charging state?

![image](https://user-images.githubusercontent.com/122752718/212742722-6984bcbb-1fc2-410e-908e-35ee2e06edc6.png)

**Comment by WillCodeForCats** [comment]
@liquidv PR #234 will add a static value for overage which will be good enough for a pre-release, but I want to do a user settable value for a full release. Once that is merged look for a pre-release you can try.

**Comment by liquidv** [comment]
Thank you, will sure to give it a try when ready
